### PR TITLE
Port posn-at-x-y

### PR DESCRIPTION
**Pull request by shaleh**

### CODE REVIEW

**Comment by shaleh** [comment]
Thoughts on `tool_bar_p` and `menu_bar_p`? Should I have made these using a `is_` form instead?

File: rust_src/src/keyboard.rs [hunk 1]
```diff
@@ -60,14 +62,52 @@ pub fn posn_at_point(pos: LispObject, window: LispObject) -> LispObject {
...
+        x += w.left_edge_x();
+
+        if whole.is_nil() {
+            x += unsafe { window_box_left_offset(w.as_ptr(), TEXT_AREA) };
+        }
+
+        y = w.frame_pixel_y(y);
+    });
+
+    return LispObject::from(unsafe {
```
Comment (Wilfred) [inline rust_src/src/keyboard.rs]: Return is unnecessary here.

File: rust_src/src/windows.rs [hunk 2]
```diff
@@ -107,6 +117,43 @@ impl LispWindowRef {
...
 
+    /// The frame x-position at which the text (or left fringe) in
+    /// window starts. This does not include a left-hand scroll bar
+    /// if any.
+    #[inline]
+    pub fn left_edge_x(self) -> i32 {
+        self.frame().as_frame_or_error().internal_border_width() + self.left_pixel_edge()
+    }
+
+    /// The frame y-position at which window starts.
```
Comment (Wilfred) [inline rust_src/src/windows.rs]: *at which the window

Likewise for the other docstrings.

File: src/keyboard.c [hunk 3]
```diff
@@ -10714,47 +10714,6 @@ The elements of this list correspond to the arguments of
...
-    IMAGE (DX . DY) (WIDTH . HEIGHT))
-The `posn-' functions access elements of such lists.  */)
-  (Lisp_Object x, Lisp_Object y, Lisp_Object frame_or_window, Lisp_Object whole)
-{
-  CHECK_NUMBER (x);
-  /* We allow X of -1, for the newline in a R2L line that overflowed
-     into the left fringe.  */
-  if (XINT (x) != -1)
-    CHECK_NATNUM (x);
-  CHECK_NATNUM (y);
```
Comment (Wilfred) [inline src/keyboard.c]: This errors if x is -2, and I don't think that's true of the Rust code.

File: src/window.c [hunk 4]
```diff
@@ -239,6 +239,31 @@ wget_pseudo_window_p(struct window *w)
...
+          && (W) == XWINDOW (WINDOW_XFRAME (W)->menu_bar_window));
+#else
+/* No menu bar windows if X toolkit is in use.  */
+  return false;
+#endif
+}
+
+/* True if W is a tool bar window.  */
+bool
+WINDOW_TOOL_BAR_P(struct window *W)
```
Comment (Wilfred) [inline src/window.c]: I think the name is fine, but I'd use lower case now they're functions.

File: src/keyboard.c (hunk 3 above)
Comment (shaleh) [reply src/keyboard.c]: Good catch.

Comment (shaleh) [reply src/keyboard.c]: I am curious how my solution will be received. I also considered and experimented with extending `LispNumber` but that  means constantly verifying the value is not a float. Which is no better.

**Comment by shaleh** [comment]
Toggle build.

**Comment by shaleh** [comment]
toggling the build

File: src/window.c (hunk 4 above)
Comment (Wilfred) [reply src/window.c]: The C code only uses upper case for macros, doesn't it?

**Comment by shaleh** [comment]
Fair points @Wilfred. New commit on its way.

**Pull request merged**